tests/cephfs/bug-2053724.py

```diff
@@ -0,0 +1,455 @@
+                            S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH))
+        {
+            if (fd == -1)
+                throw std::system_error(errno, std::system_category(),
+                buildCommittedFileError("open", directory, file, "", errno).c_str());
+        }
+
+        void WriteFd::writeAll(const void* data, size_t size)
+        {
+            size_t written(0);
+            while (written < size)
+            {
+                const ssize_t ret(::write(fd, static_cast<const char*>(data) + written, size - written));
+                if (ret < 0)
+                    /**
+                    * @todo In theory ENOSPC and EDQUOT errors could be recovered
+                    * by retrying later. If there is such retry logic at upper
+                    * layer, then we should indicate that retrying is worth it in
+                    * the exception. Currently that sounds very unlikely.
+                    */
+                    throw std::system_error(errno, std::system_category(),
+                    buildCommittedFileError("write", directory, file, "", errno).c_str());
+                else
+                    written += static_cast<size_t>(ret);
+            }
+        }
+
+        CommittedFile::CommittedFile(const std::string& filePath):
+            filePath(filePath)
+        {
+            cleanup();
+        }
+
+        CommittedFile::~CommittedFile()
+        {
+        }
+
+        void CommittedFile::write(const std::string& data)
+        {
+            DirFd dirFd(dirName(filePath));
+            /*
+            * First write and sync work-file. Do not touch real-file.
+            */
+            auto fileName(baseName(filePath));
+            auto workFileName(fileName + ".work");
+            WriteFd workFileFd(dirFd, workFileName);
+            workFileFd.writeAll(data.data(), data.size());
+            workFileFd.sync();
+            workFileFd.close();
+            /**
+            * Posix guarantees that rename is atomic...
+            */
+            dirFd.renameFile(workFileName, fileName);
+            /**
+            * ... and with a directory fsync data is actually stored on disk
+            * See: https://lwn.net/Articles/457667/
+            */
+            dirFd.sync();
+            dirFd.close();
+        }
+
+        std::string CommittedFile::read() const
+        {
+            return readFile(filePath);
+        }
+
+        void CommittedFile::cleanup()
+        {
+            /**
+            * Remove possibly existing old work file
+            */
+            const auto directory(dirName(filePath));
+            DirFd dirFd(directory);
+            const auto fileName(baseName(filePath));
+            const auto workFileName(fileName + ".work");
+            dirFd.unlink(workFileName);
+            dirFd.close();
+        }
+
+        std::string CommittedFile::getPath() const
+        {
+            return filePath;
+        }
+        """
```

Comment:
Could we store this script separately as a file or in a different repository ? 

It would allow easy maintenance and re usability.

Reply:
Okay